### PR TITLE
test: show that unverified password accounts are being linked to on social and code signups!

### PR DESCRIPTION
**Pull request by danielgent**
~#### Next~
* ~rebase when https://github.com/sesamyab/auth/pull/547is merged, I think will still fail the same!~  :x:  as expected
 
--------------------------------------------------------------------------------------
#### Simple flow: _doing on this PR_
* do sign up for an email _but not_ email verification
* sign up with social/code account with same email
* assert that not erroneously linked to unverified email account
*I'll do the fix for this here as well* :construction: :construction_worker: :building_construction: 

#### More complex permutations - _doing following on upcoming PR_
* do _another_ sign up with social/code (whichever didn't)
* verify that correctly linked to existing social/code

#### And another one
* then verify the email account and assert that linked to the primary account :upside_down_face:

### CODE REVIEW

File: test/integration/flows/magic-link-flow.spec.ts
Comment (danielgent) [inline test/integration/flows/magic-link-flow.spec.ts:1]: hit ignore whitespace and there's little additional diff - describe blocks had odd names and nestings

File: test/integration/flows/social.spec.ts [hunk 1]
```diff
@@ -638,6 +638,60 @@ describe("social sign on", () => {
...
+      const location2 = new URL(
+        socialCallbackResponse.headers.get("location")!,
+      );
+      const socialCallbackQuery2 = new URLSearchParams(location2.hash.slice(1));
+
+      const idToken = socialCallbackQuery2.get("id_token");
+      const idTokenPayload = parseJwt(idToken!);
+      expect(idTokenPayload.sub).toBe(
+        // this shows no account linking is being done to the unvalidated email account...
+        "demo-social-provider|123456789012345678901",
```
Comment (danielgent) [inline test/integration/flows/social.spec.ts:691]: :champagne: this is a bug!

File: test/integration/flows/code-flow.spec.ts [hunk 2]
```diff
@@ -1228,5 +1228,117 @@ describe("code-flow", () => {
...
+
+      const redirectUri = new URL(tokenResponse.headers.get("location")!);
+
+      const searchParams = new URLSearchParams(redirectUri.hash.slice(1));
+
+      const accessToken = searchParams.get("access_token");
+
+      const accessTokenPayload = parseJwt(accessToken!);
+      // this shows that we are linking to an unverified password account
+      expect(accessTokenPayload.sub).not.toBe(unverifiedPasswordUser._id);
```
Comment (danielgent) [inline test/integration/flows/code-flow.spec.ts:1335]: bugs! :champagne: :cockroach: :cockroach: :cockroach: :cockroach: :cockroach: :cockroach: :cockroach: :cockroach: :cockroach: 

I'll merge the primary account helper I did first but I don't think that will fix these :smile:

File: test/integration/flows/code-flow.spec.ts [hunk 3]
```diff
@@ -1228,5 +1228,117 @@ describe("code-flow", () => {
...
+
+      const idToken = searchParams.get("id_token");
+      const idTokenPayload = parseJwt(idToken!);
+      // these shows that we are linking to an unverified password account
+      expect(idTokenPayload.sub).not.toBe(unverifiedPasswordUser._id);
+      expect(idTokenPayload.email_verified).toBe(true);
```
Comment (danielgent) [inline test/integration/flows/code-flow.spec.ts:1341]: I figure we may as well start validating `email_verified`! :smi\

I could update a lot of these code tests to do this

File: src/utils/users.ts [hunk 4]
```diff
@@ -65,7 +65,19 @@ export async function getPrimaryUserByEmail({
     return;
   }
 
-  const primaryUsers = users.filter((user) => !user.linked_to);
+  const primaryUsersPossiblyWithUnverifiedAccounts = users.filter(
+    (user) => !user.linked_to,
+  );
+
+  // Should we have this ONLY check for email_verified?
```
Comment (danielgent) [inline src/utils/users.ts:72]: :eyes:  I could do that!

Then I'll make sure we're setting `email_verified: true` for code & magic link signups!  (IIRC we only ever create these when the user enters the code e.g. verifies the email)

What about social logins though? Should we actually accept a social login where the email isn't verified? *seems to me that way!* :exploding_head:

File: src/utils/users.ts [hunk 5]
```diff
@@ -65,7 +65,19 @@ export async function getPrimaryUserByEmail({
...
+  const primaryUsers = primaryUsersPossiblyWithUnverifiedAccounts.filter(
+    (user) => !(user.provider === "auth2" && !user.email_verified),
+  );
+
+  // this means the only existing account is an unverified password user
+  if (primaryUsers.length === 0) {
+    return;
+  }
```
Comment (danielgent) [inline src/utils/users.ts:80]: so we bail from this helper. tests now pass! Check the commit history to see the new tests failing before this

File: src/utils/users.ts [hunk 6]
```diff
@@ -54,12 +54,21 @@ export async function getPrimaryUserByEmail({
...
+      include_totals: false,
+      q: `email:${email}`,
+    },
+  );
+
+  // filter out unverified accounts
+  const users = usersWithUnverifiedPasswordAccounts.filter(
+    // maybe we should do this for all providers
+    (user) => !(user.provider === "auth2" && !user.email_verified),
+  );
```
Comment (danielgent) [inline src/utils/users.ts:71]: say the word and I'll change this to `(user) => user.email_verified,` and see what happens

**Comment by danielgent** [comment]
@markusahlstrand this isn't a production bug so I'll address that first :+1:

File: src/utils/users.ts (hunk 6 above)
Comment (markusahlstrand) [reply src/utils/users.ts:71]: Yes.. the question is with code logins. They may not be verified but they will be when you log in..

File: test/integration/flows/magic-link-flow.spec.ts [hunk 7]
```diff
@@ -487,75 +488,176 @@ describe("code-flow", () => {
...
+    const authenticateResponse = await client.passwordless.verify_redirect.$get(
+      {
+        query,
+      },
+    );
+    expect(authenticateResponse.status).toBe(302);
+    // ------------
+    // Authenticate using the magic link the second time
+    // ----------------
+    const authenticateResponse2 =
```
Comment (markusahlstrand) [inline test/integration/flows/magic-link-flow.spec.ts:541]: Should this really be possible? Shouldn't we only allow it to be used once?

**Review by markusahlstrand — COMMENTED** [review]
I guess we want both accounts to be verified to merge?

File: src/utils/users.ts (hunk 6 above)
Comment (danielgent) [reply src/utils/users.ts:71]: They should be, I don't think we're currently doing that. I can address that next @markusahlstrand

Comment (danielgent) [reply src/utils/users.ts:71]: Ah of course, when the user logs in we create the user and it's always set to true. 

The issue then it with social logins... Should we disallow them when `email_verified` is false?

File: test/integration/flows/code-flow.spec.ts (hunk 2 above)
Comment (danielgent) [reply test/integration/flows/code-flow.spec.ts:1551]: Fixed! Tests passing to show this works :heavy_check_mark:

**Comment by danielgent** [comment]
> I guess we want both accounts to be verified to merge?

Here https://github.com/sesamyab/auth/blob/main/test/integration/flows/password.spec.ts#L162 I'm testing 
` it("should create a new user with a password, only allow login after email validation AND link this to an existing code user with the same email", async () => {`

*but* I think you're asking for a test where we link a new code user to an existing _validated_ password user? In which case we have those on the `code-flow` integration tests. They've been there since the start actually as regression tests

File: test/integration/flows/magic-link-flow.spec.ts (hunk 7 above)
Comment (markusahlstrand) [reply test/integration/flows/magic-link-flow.spec.ts:541]: Yeah.. feel like the safer bet.

File: src/utils/users.ts (hunk 6 above)
Comment (markusahlstrand) [reply src/utils/users.ts:71]: I don't think there's a case where they are false, even though I guess it's theoretically possible

**Comment by danielgent** [comment]
I'll merge this and then address the feedback in new PRs

File: src/utils/users.ts (hunk 6 above)
Comment (danielgent) [reply src/utils/users.ts:71]: Simplifying this here then -> https://github.com/sesamyab/auth/pull/566

File: test/integration/flows/magic-link-flow.spec.ts (hunk 7 above)
Comment (danielgent) [reply test/integration/flows/magic-link-flow.spec.ts:541]: I'll start addressing this here -> https://github.com/sesamyab/auth/pull/567

**Pull request merged**